chainntnfs/height_hint_cache.go

```diff
@@ -0,0 +1,303 @@
+	spendHintBucket = []byte("spend-hints")
+
+	// confirmHintBucket is the name of the bucket which houses the height
+	// hints for transactions. Each height hint represents the earliest
+	// height at which its corresponding transaction could have been
+	// confirmed within.
+	confirmHintBucket = []byte("confirm-hints")
+
+	// ErrCorruptedHeightHintCache indicates that the on-disk bucketing
+	// structure has altered since the height hint cache instance was
+	// initialized.
+	ErrCorruptedHeightHintCache = errors.New("height hint cache has been " +
+		"corrupted")
+
+	// ErrSpendHintNotFound is an error returned when a spend hint for an
+	// outpoint was not found.
+	ErrSpendHintNotFound = errors.New("spend hint not found")
+
+	// ErrConfirmHintNotFound is an error returned when a confirm hint for a
+	// transaction was not found.
+	ErrConfirmHintNotFound = errors.New("confirm hint not found")
+)
+
+// HeightHintCache is responsible for caching the height hints for outpoints and
+// transactions. A height hint for an outpoint signifies the earliest height at
+// which this outpoint could have been spent within. A height hint for a
+// transaction signifies the earliest height at which this transaction could
+// have been included within a block.
+type HeightHintCache interface {
+	// CommitSpendHint commits a spend hint for the outpoints to the cache.
+	CommitSpendHint(heightHint uint32, ops ...wire.OutPoint) error
+
+	// QuerySpendHint returns the latest spend hint for an outpoint.
+	// ErrSpendHintNotFound is returned if a spend hint does not exist
+	// within the cache for the outpoint.
+	QuerySpendHint(op wire.OutPoint) (uint32, error)
+
+	// PurgeSpendHint removes the spend hint for the outpoints from the
+	// cache.
+	PurgeSpendHint(ops ...wire.OutPoint) error
+
+	// CommitConfirmHint commits a confirm hint for the transactions to the
+	// cache.
+	CommitConfirmHint(heightHint uint32, txids ...chainhash.Hash) error
+
+	// QueryConfirmHint returns the latest confirm hint for a transaction
+	// hash. ErrConfirmHintNotFound is returned if a confirm hint does not
+	// exist within the cache for the transaction hash.
+	QueryConfirmHint(txid chainhash.Hash) (uint32, error)
+
+	// PurgeConfirmHint removes the confirm hint for the transactions from
+	// the cache.
+	PurgeConfirmHint(txids ...chainhash.Hash) error
+}
+
+// heightHintCache is an implementation of the HeightHintCache interface backed
+// by a bolt DB instance where the hints will be stored.
+type heightHintCache struct {
+	db *bolt.DB
+}
+
+// A compile-time check to ensure our HeightHintCache implementation satisfies
+// the interface.
+var _ HeightHintCache = (*heightHintCache)(nil)
+
+// NewHeightHintCache returns a new height hint cache backed by a database.
+func NewHeightHintCache(dbDir string) (HeightHintCache, error) {
+	if err := os.MkdirAll(dbDir, 0700); err != nil {
+		return nil, err
+	}
+
+	db, err := bolt.Open(filepath.Join(dbDir, dbName), dbFilePermission, nil)
```

Comment:
Rather than create a _new_ database which will add _yet another_ memory map, this should use the existing database. Arguably, the replay log should also be within the same database.

Reply:
Fixed.